### DISCUSSION TITLE
(Testing own parser to infer grammar) What would `. tag | = "foo"` mean?

### DISCUSSION THREAD

**Discussion by amn** [mikefarah/yq#1657] [Q&A] [answered]
Mind you, the result seems different than if the white-space between the `|` and the `=` were removed. Without the white-space, I understand there's the `|=` operator, which has the effect of assigning "foo" as the value of the tag (using context of either the tag, if that even makes sense, or `.`, at the right hand side of the assignment).

With the white-space there, however, I don't understand what kind of expression I wrote, and the `-v` switch to `yq` doesn't seem to help me much, except some indication that it is no longer the equivalent of the `|=` operator being used, but rather the pipe (`|`) operator and the plain assignment (`=`) operator.

Can someone please shed some light on the difference between the two expression variants (with and without white-space), and what the one _with_ the white-space between the `|` and the `=`, means? I understand also that some explicit parentheses might make it clear, but I am also not sure about the arity of the different operators in use here. E.g. if I were to rewrite the expression as `(. tag |) = "foo"` I believe it's equivalent of the variant where the parentheses were removed (without any other change, in particular to the white-space), but that doesn't make it much more clear what kind of parse tree or expression I am looking at.

**Comment by mikefarah** [comment]
So with no space `|=` is a single operator, and the expression tree looks like:
```
- |=
   - .tag (lhs argument)
   - "foo"(rhs argument)
```

With an (invalid) expression like ".tag | = "foo"' - yq will try to build an expression tree. Now each operator has an implied precedence (kind of like a maths expression, you evaluate `*` and `/` before `+` and `-`. You can see the precedence values in lib.go for what it's worth). In this case, `=` has a higher order of precedence than `|` and so it will build an invalid expression tree like this:

```
- =
   - | (LHS argument of "=")
      -  .tag (LHS argument of "|")
   - "foo" RHS argument of "="
```
And then it discovers the "|" operator only has one argument, whereas it needs two.

**Reply by amn** [reply]
Thanks. I think I was asking specifically about the variant `. tag | = "foo"` (space between `.` and `tag`)? It isn't an invalid expression here (`yq` v4.33.3). Anyway, I suppose the syntactical correctness here generally hinges on the simple fact that both `|` and `=` are _binary_ operators -- requiring two operands, one on each side of the operator with an infix notation (that `yq` uses)?

**Comment by mikefarah** [comment, marked as answer]
Yep 👍

________________________________
